docs/getting_started/DEBUGGING.md

```diff
@@ -0,0 +1,122 @@
+# Debugging
+
+## Fetching service logs
+
+### Option A: `bolt logs`
+
+To read logs of a **single, currently running service**, run the following command:
+
+```
+bolt logs my-service
+```
+
+This command also supports `--follow` and `--stderr`.
+
+### Option B: Loki
+
+If the container has **stopped, restarted, or is a short-lived job**, you need to use Loki to read the logs.
+
+Connect to the Grafana dashboard by running the following, then visiting http://localhost:9090:
+
+```
+./scripts/forward/grafana.sh
+```
+
+Open the _Explore_ page and query `{service=~"rivet-.*"} |= "ERROR"`, or click [here](http://localhost:9090/explore?panes=%7B%22DNM%22:%7B%22datasource%22:%22loki%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bservice%3D~%5C%22rivet-.*%5C%22%7D%20%7C%3D%20%5C%22ERROR%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&schemaVersion=1&orgId=1).
+
+## Reading logs
+
+Rivet uses structured logging, which can be verbose and hard to read at times.
+
+Here is an example log:
+
+```json
+{
+	"timestamp": "2023-11-14T20:24:34.038004Z",
+	"level": "INFO",
+	"fields": {
+		"message": "operation call",
+		"body": "Request { user_ids: [Uuid(6bda49ad-355c-4f0b-b76f-cb773f4ba9df)] }"
+	},
+	"target": "rivet_operation",
+	"spans": [
+		{
+			"method": "GET",
+			"ray_id": "f14d9ff6-06bf-42dc-8939-8e228c4baa0f",
+			"uri": "/cloud/games?watch_index=1699993473573",
+			"name": "http request"
+		},
+		{ "operation": "user-get", "name": "call" }
+	]
+}
+```
+
+-   `target` is the name of the span (usually the function name) that made the log
+-   `fields` are the relevant information to the given function call
+-   `spans` are an abbreviated stack trace providing information about where the log came from
+    -   `spans[0].ray_id` is important, see below
+
+### Parsing & formatting logs
+
+Loki's [`json`](https://grafana.com/docs/loki/latest/query/log_queries/#parser-expression) parser and [`line_format`](https://grafana.com/docs/loki/latest/query/log_queries/#line-format-expression) expression are helpful for parsing structured logs.
+
+## Rays & requests
+
+Any code that runs in Rivet has an associated ray ID & request ID.
+
+A ray correlates a series of actions that started from a single event (such as an HTTP request or CRON job).
```

Comment:
Does this just mean "the idea of a ray", or is it using a ray provided by some other provider (it seems like Cloudflare has a notion of "rays"). I'm not too familiar with the term, so it might be more widly used than I know.

I'm reading a bit further, and it seems that it's just a UUID. I think it might halp to add a bit more to what a ray contains as data maybe?

Reply:
a5ba339a

Reply:
👍